src/main_cache.cpp

```diff
@@ -0,0 +1,188 @@
+#include "main_cache.hpp"
+
+MainCache::MainCache(const CACHE_SET &setting) : BaseCache(setting) {
+    property.associativity = setting.associativity;
+    property.write_policy = setting.write_policy;
+    property._num_block = (setting.cache_size << 10) / setting.block_size;
+
+    switch (property.associativity) {
+    case full_associative:
+        /* For fully associative, remaining bits are used for TAG*/
+        property._bit_index = 0;
+        property._bit_set = 0;
+        property._bit_tag = 32 - property._bit_offset;
+        break;
+    case direct_mapped:
+        property._bit_index = log2(property._num_block - 1);
+        property._bit_set = 0;
+        property._bit_tag = 32 - property._bit_offset - property._bit_index;
+        break;
+    case set_associative:
+        property._bit_index = 0;
+        property._num_way = setting.num_way;
+        property._num_set = property._num_block / property._num_way;
+        property._bit_set = log2(property._num_set - 1);
+        break;
+    }
+}
+
+MainCache::~MainCache() = default;
+
+bool MainCache::Get(const addr_t &addr) {
+    bool res = IsHit(addr);
+    if (res) {
+        _HitHandle(addr);
+    }
+    return res;
+}
+
+bool MainCache::Set(const addr_t &addr) {
+    _Replace(addr);
+    return true;
+}
+
+bool MainCache::IsHit(const addr_t &addr) {
+    ulint idx(0);
+    bool identical(true);
+
+    auto check_ident = [&](const ulint idx, const uint _bit_tag,
+                           const addr_t lhs, const addr_t rhs[]) -> bool {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Function parameter 'lhs' should be passed by const reference.](https://app.codacy.com/manual/jerry0715no14/Cache_sim/pullRequest?prid=4735708)